R/board_url.R

```diff
@@ -141,12 +179,153 @@ pin_store.pins_board_url <- function(board, name, paths, metadata,
+  "*" = "named list, where all elements are character scalars or vectors"
+)
+
+get_url_format <- function(urls) {
+  if (is_scalar_character(urls) && !is_named(urls)) {
+    return ("pins_yaml")
+  }
+  if (is_list(urls)) {
+    if (!is_named(urls) || !all(map_lgl(urls, is_character))) {
+      cli::cli_abort(
+        message = c(
+          board_url_error_message,
+          "i" = "{.var urls} resolves to a list:",
+          "i" = "- named: {.val {is_named(urls)}}",
+          "i" = "- all values character: {.val {all(map_lgl(urls, is_character))}}"
+        ),
+        class = "pins_error_board_url_argument",
+        urls = urls
+      )
+    }
+    return ("manifest_content")
+  }
+  if (is.character(urls)) {
+    if (!is_named(urls)) {
+      cli::cli_abort(
+        message = c(
+          board_url_error_message,
+          "i" = "{.var urls} resolves to a character vector, but is unnamed."
+        ),
+        class = "pins_error_board_url_argument",
+        urls = urls
+      )
+    }
+    return ("vector_of_urls")
+  }
+  cli::cli_abort(
+    message = c(
+      board_url_error_message,
+      "i" = "{.var urls} is a {.val {class(urls)}}."
+    ),
+    class = "pins_error_board_url_argument",
+    urls = urls
+  )
+}
+
+get_manifest <- function(url, call = rlang::caller_env()) {
+  # if ends with "/", look for manifest
+  if (grepl("/$", url)) {
+    url <- paste0(url, manifest_pin_yaml_filename)
+  }
+
+  # if request fails or returns with error code
+  tryCatch(
+    {
+      resp <- httr::GET(url)
+      httr::stop_for_status(resp)
+    },
+    error = function(e) {
+      cli::cli_abort(
+        message = c(
+          "Error requesting manifest file from URL {.url {url}}:",
+          " " = "{e$message}"
+        ),
+        class = "pins_error_board_url_request",
+        url = url,
+        call = call
+      )
+    }
+  )
+
+  # if file is not parsable
+  tryCatch(
+    {
+      text <- httr::content(resp, as = "text", encoding = "UTF-8")
+      manifest <- yaml::yaml.load(text)
+    },
+    error = function(e) {
+      cli::cli_abort(
+        message = c(
+          "Error parsing manifest-file at URL {.url {url}}:",
+          " " = "{e$message}",
+          "i" = "Manifest file must be text and parsable as YAML."
+        ),
+        class = "pins_error_board_url_parse",
+        resp = resp,
+        call = call
+      )
+    }
+  )
+
+  # prepend url-root to manifest entries
+  url_root <- url_dir(url)
+  manifest <- map(
+    manifest,
+    ~ map_chr(.x, ~ url_path(url_root, .x))
+  )
+
+  manifest
+}
+
+url_path <- function(url, ...) {
+  # return URL with path items appended
+  # - if last element in path ends with a slash,
+  #   ensure return URL ends with a slash
+
+  dots <- rlang::list2(...)
+  url_parsed <- httr::parse_url(url)
+  path_elems <- c(url_parsed$path, dots)
+
+  path_new <- rlang::exec(fs::path, !!!path_elems)
+  path_new <- end_with_slash(path_new)
+
+  url_parsed$path <- path_new
```

Comment:
Yeah, looking at this, I am feeling fuzzy on when this would do anything different from paste.  @ijlyttle what do you think about taking this:

```r
manifest <- map(
    manifest,
    ~ map_chr(.x, ~ url_path(url_root, .x))
  )
```

and making it this:

```r
manifest <- map(
    manifest,
    ~ map_chr(.x, ~ paste0(url_root, .x))
  )
```
For example, what happens with S3 bucket URLs?

Reply:
Just to close the loop, I made a big simplification along these lines in #686

I don't know enough about S3 URLs to answer definitively, but I think the simpler version will work so long as the directory structure is the same (which I think it is).

Reply:
Thank you! 🙌